api/graphite.go

```diff
@@ -339,53 +388,63 @@ func (s *Server) metricsIndex(ctx *middleware.Context) {
 		response.Write(ctx, response.WrapError(err))
 		return
 	}
-	errors := make([]error, 0)
-	series := make([]idx.Archive, 0)
-	seenDefs := make(map[string]struct{})
-	var mu sync.Mutex
+	reqCtx := ctx.Req.Context()
+	responses := make(chan struct {
+		series []idx.Archive
+		err    error
+	}, 1)
 	var wg sync.WaitGroup
 	for _, peer := range peers {
 		wg.Add(1)
 		if peer.IsLocal() {
 			go func() {
 				result := s.listLocal(ctx.OrgId)
-				mu.Lock()
-				for _, def := range result {
-					if _, ok := seenDefs[def.Id]; !ok {
-						series = append(series, def)
-						seenDefs[def.Id] = struct{}{}
-					}
-				}
-				mu.Unlock()
+				responses <- struct {
+					series []idx.Archive
+					err    error
+				}{result, nil}
 				wg.Done()
 			}()
 		} else {
 			go func(peer cluster.Node) {
-				result, err := s.listRemote(ctx.Req.Context(), ctx.OrgId, peer)
-				mu.Lock()
-				if err != nil {
-					errors = append(errors, err)
-				}
-				for _, def := range result {
-					if _, ok := seenDefs[def.Id]; !ok {
-						series = append(series, def)
-						seenDefs[def.Id] = struct{}{}
-					}
-				}
-				mu.Unlock()
+				result, err := s.listRemote(reqCtx, ctx.OrgId, peer)
+				responses <- struct {
+					series []idx.Archive
+					err    error
+				}{result, err}
 				wg.Done()
 			}(peer)
 		}
 	}
-	wg.Wait()
-	if len(errors) > 0 {
-		err = errors[0]
+
+	// wait for all list goroutines to end, then close our responses channel
+	go func() {
+		wg.Wait()
+		close(responses)
+	}()
```

Comment:
same as above, couldn't this routine be saved and `wg.Wait()` & `close(responses)` by called in the main thread?